### PR TITLE
Clean up entity creation logic

### PR DESCRIPTION
**Pull request by rwedge**
The idea behind this PR is to make it easier to find where in the code we handle data type and variable type conversion when creating a new entity.  Currently it's handled in two places: `EntitySet._import_from_dataframe` and `Entity.__init__`.  

This PR moves the relevant logic in `EntitySet._import_from_dataframe` into `Entity.__init__` and removes a couple statements that try to infer variable types that can be handled by `Entity.infer_variable_types` instead.

### CODE REVIEW

File: featuretools/entityset/entityset.py [hunk 1]
```diff
@@ -1142,40 +1141,6 @@ def _import_from_dataframe(self,
...
-            created_index = index
-
-        if time_index is not None and time_index not in dataframe.columns:
-            raise LookupError('Time index not found in dataframe')
-
-        for c in dataframe.columns:
-            if not is_string(c):
-                raise ValueError("All column names must be strings (Column {} is not a string)".format(c))
-
-            if dataframe[c].dtype.name.find('category') > -1:
```
Comment (kmax12) [inline featuretools/entityset/entityset.py]: in `infer_variable_types` this logic is just 

```
elif df[variable].dtype.name == "category":
      inferred_type = vtypes.Categorical
```

did this handle a case that the other didn't or were we just being extra cautious to just look for "category" anywhere in the name?

File: featuretools/entityset/entity.py [hunk 2]
```diff
@@ -62,61 +62,39 @@ def __init__(self, id, df, entityset, variable_types=None,
...
+                raise ValueError("All column names must be strings (Column {} "
+                                 "is not a string)".format(c))
+        if time_index is not None and time_index not in df.columns:
+            raise LookupError('Time index not found in dataframe')
+
+        created_index, index, df = create_index(index, make_index, df)
+        if index not in variable_types:
+            variable_types[index] = vtypes.Index
+
         self.data = {"df": df,
```
Comment (kmax12) [inline featuretools/entityset/entity.py]: can we avoid setting the data like this? perhaps by just making df an argument to some of the other functions? 

it feels wrong to set the data here and then call `self.update_data(...)` a second time at the end

File: featuretools/entityset/entity.py [hunk 3]
```diff
@@ -62,61 +62,39 @@ def __init__(self, id, df, entityset, variable_types=None,
...
-                _v = vtype[0](v, self, **vtype[1])
-            else:
-                _v = inferred_variable_types[v](v, self)
-            self.variables += [_v]
 
-        # do one last conversion of data once we've inferred
-        self.convert_all_variable_data(inferred_variable_types)
+        variable_types = variable_types or {}
+        secondary_time_index = secondary_time_index or {}
+        self.create_variables(variable_types, index, time_index, secondary_time_index)
```
Comment (kmax12) [inline featuretools/entityset/entity.py]: i think this should be `self._create_variables` since a user should never need call it, right?

Comment (kmax12) [inline featuretools/entityset/entity.py]: i think `secondary_time_index` is getting modified by reference within this function (actually when it gets passed to `infer_variable_types`). if that's true, i'd say let's avoid that or at least leave a comment.

File: featuretools/entityset/entityset.py (hunk 1 above)
Comment (rwedge) [reply featuretools/entityset/entityset.py]: I'm not sure.  It looks like from [this documentation](https://pandas.pydata.org/pandas-docs/stable/categorical.html#equality-semantics) that if we check for `df[variable].dtype == 'category'` instead we'll catch all CategoricalDtype.

Comment (kmax12) [reply featuretools/entityset/entityset.py]: cool. let's refactor to that

File: featuretools/entityset/entity.py (hunk 2 above)
Comment (rwedge) [reply featuretools/entityset/entity.py]: should we just call the relevant parts of `self.update_data` separately and skip calling `update_data`?.  It already expects the indexes to be set and it's a little strange to be "updating" the dataframe on initialization

File: featuretools/entityset/entity.py [hunk 4]
```diff
@@ -669,3 +681,21 @@ def col_is_datetime(col):
         else:
             return True
     return False
+
+
+def create_index(index, make_index, df):
```
[the remark this answers is not in the document]
Comment (rwedge) [reply featuretools/entityset/entity.py]: Maybe this should also be private

File: featuretools/entityset/entity.py (hunk 2 above)
Comment (kmax12) [reply featuretools/entityset/entity.py]: ya, i agree with that approach

File: featuretools/entityset/entity.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kmax12) [reply featuretools/entityset/entity.py]: yep

File: featuretools/entityset/entity.py (hunk 3 above)
Comment (rwedge) [reply featuretools/entityset/entity.py]: This code here, right?
```python
secondary_time_index = secondary_time_index or {}
for ti, cols in secondary_time_index.items():
    if ti not in cols:
        cols.append(ti)
```

The first line should be thrown out and mabye the rest should go in `_validate_entity_params` with either a warning or an error if the secondary time index column isn't included.

File: featuretools/entityset/entity.py [hunk 5]
```diff
@@ -359,7 +323,34 @@ def query_by_values(self, instance_vals, variable_id=None, columns=None,
...
+        # convert data once we've inferred
+        self.convert_all_variable_data(inferred_variable_types)
+        # make sure index is at the beginning
+        index_variable = [v for v in variables
+                          if v.id == index][0]
+        self.variables = [index_variable] + [v for v in variables
+                                             if v.id != index]
+
+    def infer_variable_types(self, variable_types, time_index, secondary_time_index):
         """Extracts the variables from a dataframe
```
Comment (kmax12) [inline featuretools/entityset/entity.py]: update doc string

File: featuretools/entityset/entity.py [hunk 6]
```diff
@@ -60,29 +55,32 @@ def __init__(self, id, df, entityset, variable_types=None,
...
 
         variable_types = variable_types or {}
         secondary_time_index = secondary_time_index or {}
         self._create_variables(variable_types, index, time_index, secondary_time_index)
 
         self.df = df[[v.id for v in self.variables]]
         self.set_index(index)
-        self.set_time_index(time_index, already_sorted=already_sorted)
+
+        if time_index:
```
Comment (kmax12) [inline featuretools/entityset/entity.py]: more compact
```
self.time_index = None
if time_index:
    self.set_time_index(time_index, already_sorted=already_sorted)
elif not already_sorted:
    self.df.sort_index(kind="mergesort", inplace=True)
```

**Comment by kmax12** [comment]
Looks good to merge

**Pull request merged**